### PR TITLE
SONARJS-747 Support class and property decorators

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ClassTreeImpl.java [hunk 1]
```diff
@@ -73,21 +76,26 @@ private ClassTreeImpl(
...
+    List<DecoratorTree> decorators, InternalSyntaxToken classToken, @Nullable IdentifierTreeImpl name,
     @Nullable InternalSyntaxToken extendsToken, @Nullable ExpressionTree superClass, InternalSyntaxToken openCurlyBraceToken,
     List<Tree> elements, InternalSyntaxToken closeCurlyBraceToken
   ) {
 
-    return new ClassTreeImpl(Kind.CLASS_EXPRESSION, classToken, name, extendsToken, superClass, openCurlyBraceToken, elements, closeCurlyBraceToken);
+    return new ClassTreeImpl(Kind.CLASS_EXPRESSION, decorators, classToken, name, extendsToken, superClass, openCurlyBraceToken, elements, closeCurlyBraceToken);
   }
 
   public static ClassTreeImpl newClassDeclaration(
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ClassTreeImpl.java]: MAJOR Method has 8 parameters, which is greater than 7 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ClassTreeImpl.java [hunk 2]
```diff
@@ -73,21 +76,26 @@ private ClassTreeImpl(
   }
 
   public static ClassTreeImpl newClassExpression(
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/expression/ClassTreeImpl.java]: MAJOR Method has 8 parameters, which is greater than 7 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/DecoratorTest.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  @Test
+  public void test() {
+    assertThat(Kind.DECORATOR)
+      .matches("@decorator")
+      .matches("@foo.decorator")
+      .matches("@decorator()")
+      .matches("@decorator(1, 2)")
+      .matches("@foo.decorator()")
+    ;
```
Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/DecoratorTest.java]: @vilchik-elena How about testing the recursive nature of DecoratorMemberExpression? For example:
 .matches("@foo.decorator.prop")
 .matches("@foo.decorator.prop(1, 2)")

Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/DecoratorTest.java]: @ivandalbosco fixed 077c0c09c2451fdba8cda9d1698a1c492ad062a6

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java [hunk 4]
```diff
@@ -43,6 +43,18 @@ public void with_properties() throws Exception {
...
 
+  @Test
+  public void with_decorators() throws Exception {
+    assertThat(Kind.CLASS_DECLARATION)
+      .matches("@decorator class A {}")
+      .matches("@decorator1 @decorator2 class A {}")
+      .matches("class C { @decorator method(){} }")
+      .matches("class C { @decorator static method(){} }")
+      .matches("class C { @decorator1 @decorator2 method(){} }")
+      .matches("class C {method(){} property; }")
```
Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java]: @vilchik-elena Looks like these two lines can be removed as they are cut-and-paste of method with_properties. They are irrelevant to decorators.

Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java]: @ivandalbosco thanks! removed in da165029a3882026014e73e79098b5b0e9a9b986

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java [hunk 5]
```diff
@@ -43,6 +43,18 @@ public void with_properties() throws Exception {
       .matches("class C {method(){} property; }")
       .matches("class C { property; property=42; }")
     ;
+  }
 
+  @Test
+  public void with_decorators() throws Exception {
+    assertThat(Kind.CLASS_DECLARATION)
+      .matches("@decorator class A {}")
+      .matches("@decorator1 @decorator2 class A {}")
```
Comment (ivandalbosco) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java]: @vilchik-elena In the parsing, we don't consider DecoratorList per se, we always handle it implicitly, as a collection of Decorator. Therefore we must unit test every appearance of DecoratorList. This is done in this class ClassDeclarationTest. But is missing in ClassExpressionTest. It is also missing in MethodDefinitionTest for all three cases: generator, method, accessor.

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java [hunk 6]
```diff
@@ -788,6 +789,11 @@
     CLASS_DECLARATION(ClassTree.class),
```
Comment (ivandalbosco) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java]: @vilchik-elena By the way, in JavaScriptGrammar, both CLASS_DECLARATION and CLASS_EXPRESSION produce a ClassTree. Shouldn't we rename Tree.CLASS_DECLARATION into Tree.CLASS, as it describes both cases?

Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java]: @ivandalbosco No, we want to have both kinds. E.g.
```
class A { // We know that  this is declaration
}

var B = class BB { // and this is CLASS_EXPRESSION
}
```

Comment (ivandalbosco) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java]: @vilchik-elena Sorry, I hadn't seen that Tree.CLASS_EXPRESSION exists. Thanks.

File: javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java (hunk 5 above)
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/parser/declarations/ClassDeclarationTest.java]: @ivandalbosco fixed 4bcc226067da23e4cb00c2f3f20529c4691676ec

**Pull request merged**